### PR TITLE
ck-45-Add submission errors to bulk submission response

### PR DESCRIPTION
**Pull request by theisuru**
ebi-ait/checklist#45

- Include submission errors in the response
- Change biosamples-client method signature

### CODE REVIEW

File: utils/validation/src/main/java/uk/ac/ebi/biosamples/validation/SchemaValidationService.java [hunk 1]
```diff
@@ -42,7 +42,9 @@ public SchemaValidationService(
   public Sample validate(final Sample sample) {
     final String schemaId =
         sample.getCharacteristics().stream()
-            .filter(s -> s.getType().equalsIgnoreCase("checklist"))
+            .filter(s ->
+                s.getType().equalsIgnoreCase("checklist") ||
+                    s.getType().equalsIgnoreCase("ena-checklist"))
```
Comment (dipayan1985) [inline utils/validation/src/main/java/uk/ac/ebi/biosamples/validation/SchemaValidationService.java:47]: Suggestion (not important): we have a constants class where we can keep `checklist `and` ena-checklist`.

File: webapps/core-v2/src/main/java/uk/ac/ebi/biosamples/controller/BulkActionControllerV2.java [hunk 2]
```diff
@@ -258,19 +265,36 @@ public ResponseEntity<List<Sample>> postV2(
...
-              .map(Optional::get)
-              .collect(Collectors.toList());
+      for (Sample sample : samples) {
+        Pair<Optional<Sample>, Optional<String>> sampleErrorPair =
+            persistSampleV2WebinAuth(authProvider, webinSubmissionAccountId, sample);
+        sampleErrorPair.getLeft().ifPresent(createdSamples::add);
+        sampleErrorPair.getRight().ifPresent(err -> {
+          List<SubmissionReceipt.ValidationError> validationErrors;
+          try {
+            validationErrors = objectMapper.readValue(err, new TypeReference<>() {});
```
Comment (dipayan1985) [inline webapps/core-v2/src/main/java/uk/ac/ebi/biosamples/controller/BulkActionControllerV2.java:275]: This looks good. Can you share some example error messages?

**Review by dipayan1985 — COMMENTED** [review]
Reviewed and commented.

File: webapps/core-v2/src/main/java/uk/ac/ebi/biosamples/controller/BulkActionControllerV2.java (hunk 2 above)
Comment (theisuru) [reply webapps/core-v2/src/main/java/uk/ac/ebi/biosamples/controller/BulkActionControllerV2.java:275]:
```json
{
    "samples": [
        {
            "name": "HQ171123-10",
            "accession": "SAMEA131160159",
            "sraAccession": "ERS30666820",
            "webinSubmissionAccountId": "Webin-59287",
            "taxId": 256318,
            "status": "PUBLIC",
            "release": "2023-01-03T00:29:23Z",
            "update": "2024-07-22T08:40:29.118007748Z",
            "submitted": "2024-07-22T08:40:29.118007400Z",
            "characteristics": {
                "ENA-CHECKLIST": [
                    {
                        "text": "ERC000023",
                        "tag": "attribute"
                    }
                ],
                "ENA-FIRST-PUBLIC": [
                    {
                        "text": "2023-01-03T00:29:23Z",
                        "tag": "attribute"
                    }
                ],
                "ENA-LAST-UPDATE": [
                    {
                        "text": "2023-01-03T00:29:23Z",
                        "tag": "attribute"
                    }
                ],
                "External Id": [
                    {
                        "text": "SAMEA5866940",
                        "tag": "Namespace:BioSample"
                    }
                ],
                "INSDC center name": [
                    {
                        "text": "AALBORG UNIVERSITY"
                    }
                ],
                "INSDC first public": [
                    {
                        "text": "2023-01-03T00:29:23Z"
                    }
                ],
                "INSDC last update": [
                    {
                        "text": "2023-01-03T00:29:23Z"
                    }
                ],
                "INSDC status": [
                    {
                        "text": "public"
                    }
                ],
                "SRA accession": [
                    {
                        "text": "ERS30666820"
                    }
                ],
                "Submitter Id": [
                    {
                        "text": "HQ171123-10",
                        "tag": "Namespace:AALBORG UNIVERSITY"
                    }
                ],
                "collection date": [
                    {
                        "text": "2017-02-07",
                        "tag": "attribute"
                    }
                ],
                "environment (biome)": [
                    {
                        "text": "anaerobic digester",
                        "tag": "attribute"
                    }
                ],
                "environment (feature)": [
                    {
                        "text": "anaerobic digester",
                        "tag": "attribute"
                    }
                ],
                "environment (material)": [
                    {
                        "text": "sludge",
                        "tag": "attribute"
                    }
                ],
                "geographic location (country and/or sea)": [
                    {
                        "text": "Denmark",
                        "tag": "attribute"
                    }
                ],
                "geographic location (latitude)": [
                    {
                        "text": "55.552344",
                        "unit": "DD",
                        "tag": "attribute"
                    }
                ],
                "geographic location (longitude)": [
                    {
                        "text": "9.721907",
                        "unit": "DD",
                        "tag": "attribute"
                    }
                ],
                "investigation type": [
                    {
                        "text": "metagenome",
                        "tag": "attribute"
                    }
                ],
                "organism": [
                    {
                        "text": "metagenome"
                    }
                ],
                "project name": [
                    {
                        "text": "PRJEB34014",
                        "tag": "attribute"
                    }
                ],
                "scientific_name": [
                    {
                        "text": "metagenome"
                    }
                ],
                "sequencing method": [
                    {
                        "text": "illumina",
                        "tag": "attribute"
                    }
                ],
                "title": [
                    {
                        "text": "AD-sludge-Fredericia"
                    }
                ],
                "wastewater/sludge environmental package": [
                    {
                        "text": "wastewater/sludge",
                        "tag": "attribute"
                    }
                ]
            },
            "submittedVia": "JSON_API",
            "create": "2024-07-22T08:40:29.118006693Z"
        }
    ],
    "errors": [
        {
            "sampleName": "MI.M00833_0566.001.FLD0101.b49",
            "errors": [
                {
                    "dataPath": "/characteristics/geographic location (latitude)/0/text",
                    "errors": [
                        "should match pattern \"(^[+-]?[0-9]+.?[0-9]{0,8}$)|(^not collected$)|(^not provided$)|(^restricted access$)|(^missing: control sample$)|(^missing: sample group$)|(^missing: synthetic construct$)|(^missing: lab stock$)|(^missing: third party data$)|(^missing: data agreement established pre-2023$)|(^missing: endangered species$)|(^missing: human-identifiable$)\""
                    ]
                },
                {
                    "dataPath": "/characteristics/geographic location (longitude)/0/text",
                    "errors": [
                        "should match pattern \"(^[+-]?[0-9]+.?[0-9]{0,8}$)|(^not collected$)|(^not provided$)|(^restricted access$)|(^missing: control sample$)|(^missing: sample group$)|(^missing: synthetic construct$)|(^missing: lab stock$)|(^missing: third party data$)|(^missing: data agreement established pre-2023$)|(^missing: endangered species$)|(^missing: human-identifiable$)\""
                    ]
                }
            ]
        }
    ]
}
```

**Pull request merged**